pkg/hsperfdata/hsperfdata.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package hsperfdata
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/parca-dev/parca-agent/pkg/perf"
+)
+
+const hsperfdata = "/tmp/hsperfdata_*"
+
+type Cache struct {
+	pids   map[int]struct{}
+	fs     fs.FS
+	logger log.Logger
+	mu     sync.Mutex
+	nsPID  map[int]int
+	group  singleflight.Group
+}
+
+type realfs struct{}
+
+func (f *realfs) Open(name string) (fs.File, error) {
+	return os.Open(name)
+}
+
+func NewCache(logger log.Logger) *Cache {
+	return &Cache{
+		pids:   make(map[int]struct{}),
+		fs:     &realfs{},
+		logger: logger,
+		nsPID:  map[int]int{},
+	}
+}
+
+func (c *Cache) Get(pid int) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, ok := c.pids[pid]
+	return ok
+}
+
+// IsJavaProcess returns true if the hsperfdata file exists for a given pid.
+// It first searches in all hsperfdata user directories for the processes
+// running on host and then searches in /proc/{pid}/root/tmp for processes
+// running in containers. Note that pids are assumed to be unique regardless
+// of username.
+func (c *Cache) IsJavaProcess(pid int) (bool, error) {
+	// Check if the pid is in the cache
+	if c.Get(pid) {
+		return true, nil
+	}
+
+	// Use singleflight to prevent concurrent requests for the same pid
+	ch := c.group.DoChan(strconv.Itoa(pid), func() (interface{}, error) {
+		// List all directories that match the pattern /tmp/hsperfdata_*
+		dirs, err := filepath.Glob(hsperfdata)
+		if err != nil {
+			return false, fmt.Errorf("failed to list directories: %w", err)
+		}
+
+		// Loop over all directories and search for the hsperfdata file for the given pid
+		for _, dir := range dirs {
+			hsperfdataPath := filepath.Join(dir, strconv.Itoa(pid))
+			if _, err := c.fs.Open(hsperfdataPath); err == nil {
```

Comment:
Should we close the file handles? While the Go runtime will close them when running the finalizer I think an explicit close might result in clearer code and less work for the GC

Reply:
Exactly this ☝️